scripts/js/lib/markdownTitles.ts

```diff
@@ -0,0 +1,64 @@
+import { unified } from "unified";
+import remarkParse from "remark-parse";
+import remarkGfm from "remark-gfm";
+import remarkFrontmatter from "remark-frontmatter";
+import { visit } from "unist-util-visit";
+import { Root } from "mdast";
+import yaml from "js-yaml";
+
+// Helper to recursively extract visible text from heading node
+function extractText(node: any): string {
+  if (node.type === "text" || node.type === "inlineCode") {
+    return node.value;
+  }
+
+  if (node.type === "link" && node.children) {
+    return node.children.map(extractText).join(" ");
+  }
+
+  if (node.children && Array.isArray(node.children)) {
+    return node.children.map(extractText).join(" ");
+  }
+
+  return "";
+}
+
+export async function collectHeadingTitleMismatch(
+  markdown: string,
+): Promise<string[]> {
+  const mismatches = new Set<string>();
+
+  let frontmatterTitle: string | undefined;
+  let headingText: string | undefined;
+
+  const processor = unified()
+    .use(remarkParse)
+    .use(remarkGfm)
+    .use(remarkFrontmatter, ["yaml"]);
+
+  const tree = processor.parse(markdown);
+
+  // Extract frontmatter title
+  visit(tree, "yaml", (node: any) => {
+    const data = yaml.load(node.value);
+    if (typeof data === "object" && data !== null && "title" in data) {
+      frontmatterTitle = (data as any).title;
```

Comment:
I couldn't find documentation for this, but I believe if we import `EXIT` from `unist-util-visit` then we can stop traversing the tree once we've got the information we're looking for (same for line 52):

```suggestion
      frontmatterTitle = (data as any).title;
      return EXIT;
```

Maybe see if that works and if it speeds up the check at all. Sorry I didn't think of this earlier.